### DISCUSSION TITLE
How to know is atomWithStorage loaded?

### DISCUSSION THREAD

**Discussion by ipekmuhammet** [pmndrs/jotai#1999] [Q&A] [answered]
Hi, is there anyway to know is atomWithStorage loaded from localStorage or not? for example when user logged in to app i'm storing related data in the atomWithStorage but when user closes and reopens app it shows default data = null for a second and loads to stored data after that. Because of this user sees authentication page instead of navigating directly to protected page for a 1 - 2 second. Or is there any way to trigger suspense while waiting storage to load? I didn't figure it out. Thank you.

**Comment by dai-shi** [comment, marked as answer]
Would you please try the new option `unstable_getOnInit`. Make it true and you get a different behavior.
https://github.com/pmndrs/jotai/blob/3040629485a6e247731192c67e0d518950729814/src/vanilla/utils/atomWithStorage.ts#L135

I wonder what takes one second though. Also curious, why "data=null" shows a protected page.

**Reply by mafixmafix** [reply]
thanks, it works.
but I need a stable approach in my production version. what is the best practice to store data with `atomWithStorage`? can you give me a simple example?
p.s.
when I refresh the page, first `useAtomValue` returns null and then returns the stored data. should I use some conditions to solve this or what?

**Reply by dai-shi** [reply]
If the unstable option isn't acceptable for you, a known workaround is the following:
```js
atomWithStorage('key', localStorage.getItem('key') ?? null)
```
`unstable_getOnInit` is trying to be a canonical solution for that.

**Reply by mafixmafix** [reply]
that's perfect! thanks.

**Reply by jon890** [reply]
In my case, in Next.js SSR, localStorage is undefined with SSR..
how to fix it?

**Reply by dai-shi** [reply]
If you don't specify `unstable_getOnInit`, it will use the localStorage value on the second render. So, the default behavior is tuned for SSR.

**Reply by dai-shi** [reply]
Then, you can't do SSR because the server doesn't know the localStorage value.
The known workaround is `ClientOnly` wrapper: https://github.com/pmndrs/jotai/blob/6a166125ab83e7694ca7e221e2190e15a7c6ed14/docs/utilities/storage.mdx#L54

**Reply by jon890** [reply]
Thanks. very helpful!!

**Reply by ortonomy** [reply]
How does this 'ClientOnly' wrapper affect Next's 'use client' and 'use server' - assuming that the components are not going to rendered twice on server, so you'd be totally killing the SSR by doing this for anything under the ClientOnly tree?

**Reply by dai-shi** [reply]
ClientOnly uses `useEffect` which only works in client components.

> so you'd be totally killing the SSR by doing this for anything under the ClientOnly tree?

Yes, that's the point.

**Reply by ortonomy** [reply]
@dai-shi  - I beg your pardon. Allow me to be prcise. I meant more SSR entirely. 

(conceptually, I could be completely off the mark here), but I thought the point of 'use client' was to enable some components to be client only, but if there are other non 'use client' components down the tree, they'll still be server rendered and then rehydrated into the tree as HTML.)

But with the 

```

useEffect(() => {
 setHasRendered(() => true)
}, [])


if (!hasRendered) return null
```

the entire tree will be 'null' (especially if this is top of the tree) and the server components down the tree will become client only...

it's not the same effect as simply using 'use client' ?

**Reply by dai-shi** [reply]
> but I thought the point of 'use client' was to enable some components to be client only

It sounds like there's some misunderstanding.

Regardless of Pages Router or App Router or 'use client', Nextjs's SSR renders "client components" on server. You can't read localStorage on server, so so you hit a dead end.

---
(Just a note: This is exactly why I avoid rendering client components in SSR in [Waku](https://github.com/dai-shi/waku).)

**Comment by juanhenriquez** [comment]
Hey 👋🏻! Not sure if this is the right place to ask this, but I'm facing some issues with write-only atoms that internally use atoms created with `atomWithStorage`.

Let's say we have the following atoms:
```tsx
const songsAtom = atomWithStorage("songs", []);
const playlistAtom = atomWithStorage("playlist", []);

const addSongsToPlaylistAtom = atom(null, (get, set) => {
  const songs = get(songsAtom); // <- this will always be the initial value ([]) if I don't render `songsAtom` somewhere
  set(playlistAtom, songs);
});

const saveSongsAtom = atom(null, (get, set, songs) => {
  set(songsAtom, songs);
});
```

If I call `addSongsToPlaylistAtom()`, the `get(songsAtom)` will always return the initial value of the atom (`[]` in this case) If I don't render `songsAtom` somewhere in the app.

I guess this happens because `unstable_getOnInit` is `false` by default, which makes sense for next.js applications and that's why this always return the initial data:
https://github.com/pmndrs/jotai/blob/144a6b5177e96c9ef1fd221e013c576dfdbe26a5/src/vanilla/utils/atomWithStorage.ts#L147-L150

And if I don't render the atom created with `atomWithStorage` at least once, this never gets called either:
https://github.com/pmndrs/jotai/blob/144a6b5177e96c9ef1fd221e013c576dfdbe26a5/src/vanilla/utils/atomWithStorage.ts#L157-L166


Do you guys know any workaround for this? How can I create atoms with `atomWithStorage` that are not intended to be rendered in the UI while still be able to use them inside write-only atoms and always get the latest value stored in LocalStorage?

**Reply by dai-shi** [reply]
The write function isn't reactive, so what you want is technically impossible, I guess. You could just read from the local storage, which makes sense because you never use the atom.

**Comment by smontlouis** [comment]
@dai-shi Why do we need the unstable_getOnInit ? In my naive approach `atomWithStorage` should be async, so it should suspend.
Am I missing something ?

**Reply by dai-shi** [reply]
Yeah, if it's async, unstable_getOnInit might not make much sense.
